File: src/services/mood.ts
// MoodService 前端绑定（#41，模块 I.2/I.3）— mood / energy IPC + disabled_features KV。
//
// ## 锁定项（PRD §7.9）
// - mood / energy 全 transient 不持久（line 1073/1089）：前端不缓存到 KV，仅 polling
// - disabled_features 是用户偏好，KV `pet:disabled_features` 跨重启
//
// ## IPC 与 Rust commands::mood / commands::energy 同 schema：
//   mood_get → { mood: 'neutral' | 'happy' | 'focused' | 'sleepy' | 'cozy' | 'annoyed' }
//   energy_get → { value: 0-100 }
//   mood_get_disabled_features → ['mood_icon' | 'energy' | 'free_movement', ...]
//   mood_set_disabled_features({ features: [...] })

import { invoke } from './ipc'

/** Rust Mood enum 序列化为 lowercase（commands::mood / services::mood）。 */
export type Mood = 'neutral' | 'happy' | 'focused' | 'sleepy' | 'cozy' | 'annoyed'

export interface MoodSnapshot {
  mood: Mood
}

export interface EnergySnapshot {
  value: number
}

/** disabled_features 三个有效 key（前端 UI toggle 渲染顺序）。 */
export const DISABLEABLE_FEATURES = ['mood_icon', 'energy', 'free_movement'] as const
export type DisableableFeature = (typeof DISABLEABLE_FEATURES)[number]

export async function getMood(): Promise<Mood> {
  try {
    const snapshot = await invoke<MoodSnapshot>('mood_get')
    return snapshot.mood
  } catch (e) {
    console.warn('[mood] get failed, returning neutral:', e)
    return 'neutral'
  }
}

export async function getEnergy(): Promise<number> {
  try {
    const snapshot = await invoke<EnergySnapshot>('energy_get')
    return snapshot.value
  } catch (e) {
    console.warn('[mood] energy_get failed, returning 80:', e)
    return 80
  }
}

export async function getDisabledFeatures(): Promise<DisableableFeature[]> {
  try {
    const raw = await invoke<string[]>('mood_get_disabled_features')
    return raw.filter((v): v is DisableableFeature =>
      (DISABLEABLE_FEATURES as readonly string[]).includes(v),
    )
  } catch (e) {
    console.warn('[mood] get_disabled_features failed, returning empty:', e)
    return []
  }
}

export async function setDisabledFeatures(features: DisableableFeature[]): Promise<void> {
  await invoke<void>('mood_set_disabled_features', { features })
}

/** Mood emoji 表（前端 MoodIcon 渲染用）。lowercase key 与 Rust serialize 对齐。 */
export const MOOD_EMOJI: Record<Mood, string> = {
  neutral: '',
  happy: '😺',
  focused: '🎯',
  sleepy: '😴',
  cozy: '🌙',
  annoyed: '😾',
}

/** Mood 中文 label（hover tooltip 用）。 */
export const MOOD_LABEL: Record<Mood, string> = {
  neutral: '平静',
  happy: '开心',
  focused: '专注中',
  sleepy: '困倦',
  cozy: '夜晚',
  annoyed: '不悦',
}

File: src-tauri/src/commands/mood.rs
//! Mood IPC（#41）— 单 IPC + disabled_features toggle KV 包装。
//!
//! mood 状态机详见 [services/mood.rs](../services/mood.rs)；本模块仅 IPC 端封装。
//!
//! 隐私 / 锁定项：mood 全 transient 不持久（PRD line 1089）；本 IPC 仅读，不接受外部写入。
//! disabled_features 是用户偏好（持久 KV `pet:disabled_features`），与 mood 数据本身分层。

use serde::Serialize;
use tauri::{AppHandle, Runtime, State};

use crate::services::mood::{Mood, MoodState};

#[derive(Serialize)]
pub struct MoodSnapshot {
    pub mood: Mood,
}

/// 读当前 mood（含 transient 合并后的"展示值"）。前端 MoodIcon 浮层 polling 用。
#[tauri::command]
pub fn mood_get(state: State<'_, MoodState>) -> MoodSnapshot {
    MoodSnapshot {
        mood: state.compute_current(),
    }
}

/// disabled_features KV 读写。值是 JSON 数组（["mood_icon","energy","free_movement"] 子集）。
/// 跨重启持久（PRD §7.9.5 验收 4 "各 feature 独立可关"）。
const DISABLED_FEATURES_KEY: &str = "pet:disabled_features";

#[tauri::command]
pub async fn mood_get_disabled_features<R: Runtime>(app: AppHandle<R>) -> Result<Vec<String>, String> {
    match crate::services::config::get(&app, DISABLED_FEATURES_KEY).await {
        Ok(Some(raw)) => serde_json::from_str(&raw).map_err(|e| {
            format!(
                "disabled_features JSON parse failed (returning empty list): {e}"
            )
        }),
        Ok(None) => Ok(vec![]),
        Err(e) => Err(e.to_string()),
    }
}

#[tauri::command]
pub async fn mood_set_disabled_features<R: Runtime>(
    app: AppHandle<R>,
    features: Vec<String>,
) -> Result<(), String> {
    let json = serde_json::to_string(&features)
        .map_err(|e| format!("serialize disabled_features failed: {e}"))?;
    crate::services::config::set(&app, DISABLED_FEATURES_KEY, &json)
        .await
        .map_err(|e| e.to_string())
}
